File: src/Tests/Behavioral/GoCallVariations/GoCallVariations.go
package main

import (
	"fmt"
	"time"
)

func main() {
	go fmt.Println("First")
	go fmt.Println("Second")
	go fmt.Println("Third")

	f1 := fmt.Println
	go f1("Fourth")

	go GetPrintLn()("Fifth")

	go fmt.Println("Function result:", add(3, 4))

	printSquare(5)

	count := 1
	go func() {
		fmt.Println("Go count (closure):", count)
	}()
	count = 10
	fmt.Println("Count before Go:", count)

	time.Sleep(200)

	done := make(chan struct{})
	runPair(done)
	<-done

	fmt.Println("Main function")
}

func GetPrintLn() func(string) {
	return func(src string) {
		fmt.Println(src)
	}
}

func add(x, y int) int {
	result := x + y
	fmt.Println("Calculate:", result)
	return result
}

// runPair mirrors net lookup.go's `go dnsWaitGroupDone(ch, func() {})`: a FUNC-VALUE
// callee with a func-literal ARGUMENT — the literal's capture-snapshot declarations must
// hoist BEFORE the goǃ statement, not into the argument list (CS1003 x4).
func runPair(done chan struct{}) {
	tag := "pair"
	handler := func(ch chan struct{}, fn func()) {
		fn()
		fmt.Println("handled:", tag)
		ch <- struct{}{}
	}
	go handler(done, func() { fmt.Println("inner fn ran") })
}

func printSquare(n int) {
	go fmt.Println("Go thread square:", n*n)
	n++
	fmt.Println("Immediate n:", n)
}

File: src/Tests/Behavioral/GoCallVariations/GoCallVariations.cs
namespace go;

using fmt = fmt_package;
using time = time_package;

partial class main_package {

internal static void Main() {
    goǃ(ᴛ1 => fmt.Println(ᴛ1), "First");
    goǃ(ᴛ1 => fmt.Println(ᴛ1), "Second");
    goǃ(ᴛ1 => fmt.Println(ᴛ1), "Third");
    var f1 = fmt.Println;
    var f1ʗ1 = f1;
    goǃ(ᴛ1 => f1ʗ1(ᴛ1), "Fourth");
    goǃ(GetPrintLn(), (@string)"Fifth");
    goǃ((ᴛ1, ᴛ2) => fmt.Println(ᴛ1, ᴛ2), "Function result:", add(3, 4));
    printSquare(5);
    nint count = 1;
    goǃ(() => {
        fmt.Println("Go count (closure):", count);
    });
    count = 10;
    fmt.Println("Count before Go:", count);
    time.Sleep(200);
    var done = new channel<EmptyStruct>(1);
    runPair(done);
    ᐸꟷ(done);
    fmt.Println("Main function");
}

public static Action<@string> GetPrintLn() {
    return (@string src) => {
        fmt.Println(src);
    };
}

internal static nint add(nint x, nint y) {
    nint result = x + y;
    fmt.Println("Calculate:", result);
    return result;
}

internal static void runPair(channel<EmptyStruct> done) {
    @string tag = "pair"u8;
    var handler = (channel<EmptyStruct> ch, Action fn) => {
        fn();
        fmt.Println("handled:", tag);
        ch.ᐸꟷ(new EmptyStruct());
    };
    var handlerʗ1 = handler;
    goǃ(handlerʗ1, done, () => {
        fmt.Println("inner fn ran");
    });
}

internal static void printSquare(nint n) {
    goǃ((ᴛ1, ᴛ2) => fmt.Println(ᴛ1, ᴛ2), "Go thread square:", n * n);
    n++;
    fmt.Println("Immediate n:", n);
}

} // end main_package
